datastore/emulator.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datastore
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"time"
+
+	//"net/http"
+	"os"
+	"os/exec"
+	"syscall"
+)
+
+// Fake projectID set by the emulator
+var fakeProjectID = "testapp"
+
+// Port for emulator to use
+var emulatorPort = 30801
+
+// Datastore emulator for use in tests
+//
+// Background: this was devised due to aetest package no longer being available for 1.12+ runtimes as dev_appserver.py
+// cannot run runtimes past go111, see: https://github.com/googleapis/google-cloud-go/issues/644
+//
+//   Usage:
+//
+//	 ctx, cancel := context.WithCancel(context.Background())
+//	 // Save the emulator's quit channel
+//	 quit := Emulator(ctx)
+//	 code := m.Run()
+//	 cancel()
+//	 // Wait for channel close
+//	 <-quit
+//
+func Emulator(ctx context.Context) <-chan bool {
+
+	// Set log formatter to process newlines since emulator gives a lot of output
+	log.SetFormatter(&log.TextFormatter{DisableQuote: true})
+
+	// Set up the command
+	var out bytes.Buffer
+	// --in-memory flag is important as DS emulator doesn't allow reset if data is stored on disk
+	cmd := exec.Command("gcloud",
+		"beta", "emulators", "datastore", "start",
+		fmt.Sprintf("--host-port=0.0.0.0:%v", emulatorPort),
+		"--no-store-on-disk",       // use in-memory store
+		"--consistency=1.0",        // disable random failure injection
+		"--project="+fakeProjectID, // use a fake project id so it doesn't try to infer one from environment
+	)
+	cmd.Stdout = &out
+	cmd.Stderr = &out
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true} // make this process a new group leader to help cleanup
+
+	quit := make(chan bool)
+
+	log.Info("booting up datastore emulator...")
+	err := cmd.Start()
+	time.Sleep(2 * time.Second)
+	if err != nil {
+		log.Fatalf("unable to start DS emulator: %v\n%v", err, out.String())
+		os.Exit(1)
+	}
+
+	ch := make(chan struct{})
+
+	go func() {
+		defer close(ch)
+		err := cmd.Wait()
+		if err != nil {
+			log.Infof("Datastore emulator killed: %v\n\t%v", err, out.String())
+		}
+		log.Info("Datastore emulator exited.")
+		quit <- true
+	}()
+
+	go func() {
+		select {
+		// Stop datastore emulator goroutine when context is finished
+		case <-ctx.Done():
+			cmd.Process.Signal(os.Interrupt)
+			// Give emulator time to shut down
+			time.Sleep(2 * time.Second)
+			quit <- true
+
+		case <-ch:
```

Comment:
You can change this to wait on the `quit` channel. That way you can remove the `ch` channel completely. You don't need 2 channels for the same notification.

Reply:
Agreed. Done.